subsys/logging/log_backend_cavs_hda.c

```diff
@@ -0,0 +1,247 @@
+
+	memcpy(&hda_log_buf[idx], data, copy_len);
+
+	/* There may be a wrapped copy */
+	size_t wrap_copy_len = length - copy_len;
+
+	if (wrap_copy_len != 0) {
+		memcpy(&hda_log_buf[0], &data[copy_len], wrap_copy_len);
+	}
+
+	ret = length;
+	hda_log_buffered += length;
+
+exit:
+	atomic_set(&hda_log_lock, 0);
+
+	if (hda_log_buffered > sizeof(hda_log_buf)/2) {
+		hda_log_flush();
+	}
+
+	return ret;
+}
+/**
+ * 128 bytes is the smallest transferrable size for HDA so use that
+ * and encompass almost all log lines in the formatter before flushing
+ * and memcpy'ing to the HDA buffer
+ */
+#define LOG_BUF_SIZE 128
+static uint8_t log_buf[LOG_BUF_SIZE];
+LOG_OUTPUT_DEFINE(log_output_cavs_hda, hda_log_out, log_buf, LOG_BUF_SIZE);
+
+static void hda_log_periodic(struct k_timer *tm)
+{
+	ARG_UNUSED(tm);
+
+	hda_log_flush();
+}
+
+static inline void dropped(const struct log_backend *const backend,
+			   uint32_t cnt)
+{
+	ARG_UNUSED(backend);
+
+	log_output_dropped_process(&log_output_cavs_hda, cnt);
+}
+
+static void panic(struct log_backend const *const backend)
+{
+	ARG_UNUSED(backend);
+
+	log_backend_std_panic(&log_output_cavs_hda);
+}
+
+static int format_set(const struct log_backend *const backend, uint32_t log_type)
+{
+	ARG_UNUSED(backend);
+
+	log_format_current = log_type;
+	return 0;
+}
+
+static uint32_t format_flags(void)
+{
+	uint32_t flags = LOG_OUTPUT_FLAG_LEVEL | LOG_OUTPUT_FLAG_TIMESTAMP;
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_FORMAT_TIMESTAMP)) {
+		flags |= LOG_OUTPUT_FLAG_FORMAT_TIMESTAMP;
+	}
+	return flags;
+}
+
+static volatile uint32_t counter;
+
+static void process(const struct log_backend *const backend,
+		union log_msg2_generic *msg)
+{
+	ARG_UNUSED(backend);
+
+	log_format_func_t log_output_func = log_format_func_t_get(log_format_current);
+
+	log_output_func(&log_output_cavs_hda, &msg->log, format_flags());
+}
+
+static void init(const struct log_backend *const backend)
+{
+	ARG_UNUSED(backend);
+
+	int res;
+	uint32_t channel = CONFIG_LOG_BACKEND_CAVS_HDA_CHANNEL;
+
+	hda_log_dev = device_get_binding("HDA_HOST_IN");
+	__ASSERT(hda_log_dev, "No valid DMA device found");
+
+	hda_log_chan = dma_request_channel(hda_log_dev, &channel);
+	__ASSERT(hda_log_chan >= 0, "No valid DMA channel");
+	__ASSERT(hda_log_chan == channel, "Not requested channel");
+
+	hda_log_buffered = 0;
+
+	/* configure channel */
+	struct dma_block_config hda_log_dma_blk_cfg = {
+		.block_size = CONFIG_LOG_BACKEND_CAVS_HDA_SIZE,
+		.source_address = (uint32_t)(uintptr_t)&hda_log_buf,
+	};
+
+	struct dma_config hda_log_dma_cfg = {
+		.channel_direction = MEMORY_TO_HOST,
+		.block_count = 1,
+		.head_block = &hda_log_dma_blk_cfg,
+		.source_data_size = 4,
+	};
+
+	res = dma_config(hda_log_dev, hda_log_chan, &hda_log_dma_cfg);
+	__ASSERT(res == 0, "DMA config failed");
+
+	res = dma_start(hda_log_dev, hda_log_chan);
+	__ASSERT(res == 0, "DMA start failed");
+
+	k_timer_init(&hda_log_timer, hda_log_periodic, NULL);
```

Comment:
do we need this here, shouldn't this be handled by the core logging infrastructure? @dcpleung  can you please confirm?

Reply:
The logging subsys doesn't have an explicit way of forcing a transfer. The `log_output_flush()` simply calls the backend's output function to send the partially filled buffer to the output. There is no "end of message' notification to the backend. If there is a large enough buffer, you can wait till the buffer is filled with one log message, and start the DMA transfer when the backend output function is called. However, in logging deferred mode, multiple log messages may be processed at the same time, and you will have a series of DMA transfer requests.

Reply:
Right so the way this is done here with a watermark and timer to ensure a periodic flush (when the watermark doesn't get hit) makes sense then? I can imagine maybe in panic mode I should always flush perhaps then, with some 0 byte padding maybe? I was afraid of padding the output because I wasn't sure if 0 byte padding sys-t dictionary or sys-t formatting would screw things up, but that would be needed to get immediate dma writes all the time as there's a minimum 128 byte transfer

Reply:
Padding would probably not work with dictionary and sys-t catalog messages. I think you can do a DMA transfer on a partially filled block. Let the next log message fill the rest of the block, and do another transfer of the same block. Though it would require the log streamer on the other side to be aware of this.

Reply:
Just to pile on from a comment elsewhere (that I never pushed, sorry).  I don't think the timer is doing much good here.  Let the app decide on flushing policy (creating a period timer to call flush() in the app is really trivial).  And if we do want a timer here, it should probably leverage its better access for more intelligence: have reset itself when we write to a an empty-and-idle buffer, such that the budget becomes the "most time a byte of output will sit unflushed" and we promise no spurious/tiny output except when really needed.

Reply:
This is coming directly from the way dma-trace works in sof, https://github.com/thesofproject/sof/blob/main/src/trace/dma-trace.c#L492